### PR TITLE
Introduce macros to use reflex attributes

### PR DESCRIPTION
**Pull request by fwyzard**
#### PR description:

The way dictionary informations are propagated from the C++ code or XML dictionaries to reflex and cling is rather roundabout:
  - `<field name="data_" comment="!"/>` tags in XML dictionaries are parsed by `genreflex` and injected into the LLVM AST of the corresponding C++ code as comments `//!`;
  - C++ comments like `//!` or `//[size_]` are converted by `genreflex`/`rootcling` into LLVM AST annotations;
  - `cling` parses the LLVM annotations and uses them to generate the desired behaviour in the dictionaries.

This approach does not work well with macro-generated data members:
  - macros cannot generate comments, so `//!` or `//[size_]` cannot be used directly;
  - macros cannot easily be used to generate the `class_def.xml` file, requiring manual intervention for their implementation and maintenance.

However, it turns out that dictionaries can bypass the comments and use LLVM annotations directly within the C++ code. So
```c++
private:
  int size_;
  float* data_;       //[size_]
  float* transient_;  //!
```
can be also expressed as
```c++
private:
  int size_;
  float* data_ [[clang::annotate("[size_]")]];
  float* transient_ [[clang::annotate("!")]];
```
and annotations _can_ be generated by macros.

A new header file, `FWCore/Reflection/interface/reflex.h`, currently implements two macros:
  - `EDM_REFLEX_TRANSIENT` can be used to annotate transient data members, like `//!`
  - `EDM_REFLEX_SIZE(SIZE)` can be used to annotate dynamic arrays, like `//[SIZE]`

To avoid warning about unrecognised attributes, these macros expand to nothing unless `__CLING__` is defined.

The changes to `DataFormats/SoATemplate/interface/SoALayout.h` add these macros to the SoA data members, so they do not need to be explicitly added by hand to the various `classes_def.xml` files.

#### PR validation:

The SoA unit tests under `HeterogeneousCore/AlpakaTest` and `HeterogeneousCore/CUDATest` pass.

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

No backport expected.

### CODE REVIEW

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
@ericcano FYI

**Comment by fwyzard** [comment]
@makortel @Dr15Jones @pcanal what do you think ?

**Comment by makortel** [comment]
I think the annotations of class member variables in the class definition is a better way to express the information than using a separate file (`classes_def.xml`) or "magic comments". Thanks for finding it out!

@pcanal Can you think of any (potential) downsides from the ROOT point of view?

**Comment by makortel** [comment]
Comparison differences are in 11634.7 (#39803)

**Comment by fwyzard** [comment]
It does work right now, and of course I do not mind making changes to the spelling if needed, or moving t centrally supplied macros.

One thing to note, though: I understand that the current version of cling is based on clang 9; more recent versions use (or at least support) a more general syntax for annotations.
So if cling moves to a more recent code base in the future, we may need to update the definition of the macros.

**Comment by makortel** [comment]
> So if cling moves to a more recent code base in the future, we may need to update the definition of the macros.

ROOT 6.28 will come with LLVM 13 (and we already have that in the ROOT6_X IBs).

**Comment by fwyzard** [comment]
Can we trigger the tests for the ROOT6_X IB ?

**Comment by makortel** [comment]
@cmsbuild, please test for CMSSW_13_0_ROOT6_X

**Comment by fwyzard** [comment]
OK, so the clang 13-based version of cling works, too.

**Comment by fwyzard** [comment]
+heterogeneous

**Comment by makortel** [comment]
@fwyzard Can you confirm that the test `testHeterogeneousCoreAlpakaTestWriteRead` would fail if either of the `EDM_REFLEX_TRANSIENT` or `EDM_REFLEX_SIZE` annotations would not work properly?

@Dr15Jones Do you have any comments about the nomenclature?

File: FWCore/Utilities/interface/reflex.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
+#ifndef FWCore_Utilities_interface_reflex_h
```
Comment (Dr15Jones) [inline FWCore/Utilities/interface/reflex.h:1]: I wonder if this more naturally fits in FWCore/Reflection?

Comment (fwyzard) [reply FWCore/Utilities/interface/reflex.h:1]: Ehm, so, should I move them or not ?

**Comment by fwyzard** [comment]
> @fwyzard Can you confirm that the test `testHeterogeneousCoreAlpakaTestWriteRead` would fail if either of the `EDM_REFLEX_TRANSIENT` or `EDM_REFLEX_SIZE` annotations would not work properly?

Yes, if the macros did not work we would get an error like
```
----- Begin Fatal Exception 09-Jan-2023 23:20:27 CET-----------------------
An exception of category 'FatalRootError' occurred while
   [0] Calling EventProcessor::runToCompletion (which does almost everything after beginJob and before endJob)
   Additional Info:
      [a] Fatal Root Error: @SUB=TStreamerInfo::Build
portabletest::TestSoALayout<128,false>, discarding: byte* mem_, no [dimension]


----- End Fatal Exception -------------------------------------------------
```

File: FWCore/Utilities/interface/reflex.h (hunk 1 above)
Comment (makortel) [reply FWCore/Utilities/interface/reflex.h:1]: `FWCore/Reflection` depends only on `rootcling` (in addition of `FWCore/Utilities`), of which every package generating dictionaries should already depend anyway. I see `DataFormats/SoATemplate` itself does not depend on ROOT in any way, but I'd expect every use of it in conjunction with `DataFormats/Portable` to depend on 'rootcling` at minimum.

Technically this header file does not depend on ROOT, but conceptually it does (it is meant to be used in conjunction with `rootcling`), and the functionality is related to `FWCore/Reflection`  E.g. `DataFormats/Provenance` depends on `FWCore/Reflection`, so I'd expect the practical dependence impact to be minimal.

I'd be in favor of moving the header to `FWCore/Reflection`.

**Comment by fwyzard** [comment]
+heterogeneous

**Comment by fwyzard** [comment]
The only failed unit test I'd say is unrelated to these changes:
```
===== Test "test-das-selected-lumis" ====
+ export CMS_BOT_USE_DASGOCLIENT=true
+ CMS_BOT_USE_DASGOCLIENT=true
+ QUERY='lumi,file dataset=/HIHardProbes/HIRun2018A-v1/RAW run=326479'
+ grep '^/store/'
+ das-selected-lumis.py 1,23
+ dasgoclient --limit 0 --query 'lumi,file dataset=/HIHardProbes/HIRun2018A-v1/RAW run=326479' --format json
2023/01/10 17:56:28 ERROR failed to parse X509 proxy: crypto/tls: failed to parse key: asn1: syntax error: sequence truncated

---> test test-das-selected-lumis had ERRORS
TestTime:1
^^^^ End Test test-das-selected-lumis ^^^^
```

**Comment by makortel** [comment]
+core

Comparison failures are in 11634.7 (#39803)

**Comment by Axel-Naumann** [comment]
Please note that the annotations used by rootcling / genreflex are currently an undocumented implementation detail and not part of ROOT's public API.

While https://github.com/root-project/root/issues/12012 suggests to change that, I am fairly certain that the spelling of the attributes will change in the process. We are happy to work with you on a solution! Shall we discuss in the ROOT issue?

How urgent / important is this feature?

**Comment by fwyzard** [comment]
Hi @Axel-Naumann 

> Please note that the annotations used by rootcling / genreflex are currently an undocumented implementation detail and not part of ROOT's public API.

I know, it took some digging to figure it out :-D


> I am fairly certain that the spelling of the attributes will change in the process. 

No problem, it should be trivial to update the definition of the macros accordingly.

> We are happy to work with you on a solution! Shall we discuss in the ROOT issue?

:+1:

> How urgent / important is this feature?

It's... going in production for data taking in a couple months time :-)

**Comment by Axel-Naumann** [comment]
> It's... going in production for data taking in a couple months time :-)

Well okay, for all means go for it :-) but that also means that making this an official interface isn't as urgent? "Just" important?

**Comment by fwyzard** [comment]
> Well okay, for all means go for it :-) but that also means that making this an official interface isn't as urgent? "Just" important?

👍🏻

**Comment by rappoccio** [comment]
+1

Errors are unrelated to this PR

**Comment by makortel** [comment]
@rappoccio I believe this PR would need the explicit `merge` given the "failed" test report.

**Comment by perrotta** [comment]
merge

**Pull request merged**